From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 20 Apr 2026 01:10:30 +0800
Subject: [PATCH] Fix entity portal-teleport speed


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 35ab432bbd59aba80a6eaca1769e6aa82c8a7e46..1eddf666463da5331d8a6c177291fb0c7833160d 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1492,7 +1492,31 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         foliaProfiler.startTimer(timerId);
         try {
         // Folia end - profiler
+        // Lophine start - Entity portal-teleport speed fix
         if (isActive) { // Paper - EAR 2
+            if (!(entity instanceof Player) && entity.teleportTickType == fun.bm.lophine.enums.EnumTeleportStage.NEED_SYNC) { // Luminol - after portal compensate tick
+                entity.tick();
+                while (entity.losingTicks-- > 0) {
+                    entity.tick();
+                }
+                entity.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.NORMAL;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else if (!(entity instanceof Player) && entity.teleportTickType == fun.bm.lophine.enums.EnumTeleportStage.STOP_TICKING) { // Luminol - portal teleport only
+                if (fun.bm.lophine.config.modules.fixes.TeleportConfig.enableDelayCompensation && entity.losingTicks < fun.bm.lophine.config.modules.fixes.TeleportConfig.maxDelayCompensationTicks) {
+                    entity.losingTicks++;
+                }
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else {
         entity.tick();
         // Folia start - region threading
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
@@ -1503,6 +1527,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             // portalled
             return;
         }
+        }
+        // Lophine end - Entity portal-teleport speed fix
         // Folia end - region threading
         } else {entity.inactiveTick();} // Paper - EAR 2
         profiler.pop();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 6fa3ac9a37dafa166d8e70aaf4ec369db0792200..07b2d9ab0dd5bee9852fede2194dc4f3571d258e 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -388,6 +388,8 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public fun.bm.lophine.enums.EnumTeleportStage teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.NORMAL; // Lophine - Entity portal-teleport speed fix
+    public int losingTicks = 0; // Lophine - Entity portal-teleport speed fix
 
     public void inactiveTick() {
     }
@@ -3597,6 +3599,7 @@ public abstract class Entity
         } else {
             if (this.portalProcess == null || !this.portalProcess.isSamePortal(portal)) {
                 this.portalProcess = new PortalProcessor(portal, pos.immutable());
+                this.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.STOP_TICKING; // Lophine - Entity portal-teleport speed fix
             } else if (!this.portalProcess.isInsidePortalThisTick()) {
                 this.portalProcess.updateEntryPosition(pos.immutable());
                 this.portalProcess.setAsInsidePortalThisTick(true);
@@ -4575,6 +4578,7 @@ public abstract class Entity
 
     public void postChangeDimension() {
         this.resetStoredPositions();
+        this.teleportTickType = fun.bm.lophine.enums.EnumTeleportStage.NEED_SYNC; // Lophine - Entity portal-teleport speed fix
     }
 
     protected static enum PortalType {
