From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 8 Jun 2025 01:23:55 +0800
Subject: [PATCH] Rebrand to Lophine


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 4e099b584c3435ba1ebce492011d35ae1ac0cdf8..378c78b9321cbb96d95f0fb0d8c9c2ca2a283752 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -17,6 +17,17 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    // Lophine start
+    /**
+     * The brand id for folia.
+     */
+    Key BRAND_FOLIA_ID = Key.key("papermc", "folia");
+    /**
+     * The brand id for Lophine.
+     */
+    Key BRAND_LOPHINE_ID = Key.key("lophinelabs", "lophine");
+    // Lophine end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
